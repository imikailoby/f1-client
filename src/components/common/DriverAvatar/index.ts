export { default as DriverAvatar } from './driverAvatar.component';
