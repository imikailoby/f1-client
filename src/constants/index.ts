export const API_URL: string = process.env.REACT_APP_API_URL || 'http://ergast.com/api/f1';
export const CURRENT_ENV: 'development' | 'production' | 'test' = process.env.NODE_ENV as
  | 'development'
  | 'production'
  | 'test';

// Value hardcoded because we cannot time traveling and start first F1 season before 1950;
export const FIRST_F1_SEASON_YEAR = 1950;
export const FROM_F1_SEASON_YEAR = Number(process.env.REACT_APP_FROM_F1_SEASON_YEAR || '2005');

export const WIKI_API_URL = 'https://en.wikipedia.org/w/api.php';
