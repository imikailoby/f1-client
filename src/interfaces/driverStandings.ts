import { Constructor } from './constructor';
import { Driver } from './driver';

export interface DriverStandingsItem {
  Constructors: Constructor[];
  Driver: Driver,
  points: string;
  positionText: string;
  wins: string;
}
