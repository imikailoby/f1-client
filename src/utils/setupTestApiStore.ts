/* eslint-disable @typescript-eslint/no-explicit-any */
import { AnyAction, configureStore, EnhancedStore, Middleware, Reducer } from '@reduxjs/toolkit';

export function setupTestApiStore<
  A extends {
    reducer: Reducer<any, any>;
    reducerPath: string;
    middleware: Middleware;
    util: { resetApiState(): any };
  },
>(api: A) {
  const getStore = () =>
    configureStore({
      reducer: {
        [api.reducerPath]: api.reducer,
      },
      middleware: (gdm) => gdm({ serializableCheck: false, immutableCheck: false }).concat(api.middleware),
    });

  type StoreType = EnhancedStore<
    {
      api: ReturnType<A['reducer']>;
    },
    AnyAction,
    ReturnType<typeof getStore> extends EnhancedStore<any, any, infer M> ? M : never
  >;

  const initialStore = getStore() as StoreType;
  const refObj = {
    api,
    store: initialStore,
  };

  return refObj;
}
