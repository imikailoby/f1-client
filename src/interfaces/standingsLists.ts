import { DriverStandingsItem } from './driverStandings';

export interface StandingsListItem {
  season: string;
  round: string;
  DriverStandings: [DriverStandingsItem];
}
