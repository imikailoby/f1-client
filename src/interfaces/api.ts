import { MRData } from '../types/mrData';
import { RaceTable } from './raceTable';
import { StandingsListItem } from './standingsLists';

export interface IGetDriverStandingsResponse {
  MRData: MRData<{ StandingsTable: { StandingsLists: StandingsListItem[] } }>
}

export interface IGetResultsByYearResponse {
  MRData: MRData<{ RaceTable: RaceTable }>
}

export interface IGetDriverStandings {
  standingsLists: StandingsListItem[],
  pagination: {
    limit: MRData['limit'];
    offset: MRData['offset'];
    total: MRData['total'];
  }
}
