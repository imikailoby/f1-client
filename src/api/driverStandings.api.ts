import { createApi, fetchBaseQuery } from '@reduxjs/toolkit/query/react';
import { IGetDriverStandings, IGetDriverStandingsResponse } from '../interfaces/api';
import { API_URL } from '../constants';
import { transformGetDriverStandingsApiResponse } from '../helpers/api.helpers';

export const driverStandingsApi = createApi({
  reducerPath: 'driverStandings',
  baseQuery: fetchBaseQuery({ baseUrl: API_URL }),
  tagTypes: ['DriverStandings'],
  endpoints: (builder) => ({
    getDriverStandings: builder.query<IGetDriverStandings, { limit: number, offset: number }>({
      query: (params) => ({ url: 'driverstandings/1.json', params }),
      providesTags: ['DriverStandings'],
      transformResponse: (response: IGetDriverStandingsResponse) => transformGetDriverStandingsApiResponse(response),
    }),
  }),
});

export const { useGetDriverStandingsQuery } = driverStandingsApi;
