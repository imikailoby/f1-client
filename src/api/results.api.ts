import { createApi, fetchBaseQuery } from '@reduxjs/toolkit/query/react';
import { IGetResultsBySeasonResponse } from '../interfaces/api';
import { API_URL } from '../constants';
import { RaceTable } from '../interfaces/raceTable';
import { transformGetResultsBySeasonResponse } from '../helpers/api.helpers';
import { StandingsListItem } from '../interfaces/standingsLists';

export const resultsApi = createApi({
  reducerPath: 'results',
  baseQuery: fetchBaseQuery({ baseUrl: API_URL }),
  tagTypes: ['Results'],
  endpoints: (builder) => ({
    getResultsBySeason: builder.query<RaceTable, { season: StandingsListItem['season'] }>({
      query: ({ season }) => `${season}/results/1.json`,
      providesTags: ['Results'],
      transformResponse: (response: IGetResultsBySeasonResponse) => transformGetResultsBySeasonResponse(response),
    }),
  }),
});

export const { useGetResultsBySeasonQuery } = resultsApi;
