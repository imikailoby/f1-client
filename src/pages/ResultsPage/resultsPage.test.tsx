/**
 * @jest-environment jsdom
 */
import { rest } from 'msw';
import { render, screen, act } from '@testing-library/react';
import { Provider } from 'react-redux';
import { server } from '../../../mocks/server';
import { resultsApi } from '../../api/results.api';
import getResultsBySeasonResponse from '../../../mocks/responses/getResultsBySeasonResponse.json';
import getDriverStandingsReponse from '../../../mocks/responses/getDriverStandingsReponse.json';
import { API_URL, WIKI_API_URL } from '../../constants';
import { ResultsPage } from '.';
import { clearSessionStorage } from '../../utils/common';
import { StandingsListItem } from '../../interfaces/standingsLists';
import { setupTestApiStore } from '../../utils/setupTestApiStore';

const storeRef = setupTestApiStore(resultsApi);

const MOCKED_DATA = {
  ...getDriverStandingsReponse.MRData.StandingsTable.StandingsLists[0],
  season: '2005',
} as unknown as StandingsListItem;

beforeAll(() => server.listen());
afterEach(() => server.resetHandlers());
afterAll(() => {
  server.close();
  clearSessionStorage();
});

describe('<ResultsPage />', () => {
  test('renders with loading state without crashing', async () => {
    server.use(
      rest.get(`${API_URL}/${MOCKED_DATA.season}/results/1.json`, (_, res, ctx) => res(ctx.status(500))),
      rest.get(`${WIKI_API_URL}*`, (_, res, ctx) => res(ctx.status(500))),
    );

    const { container } = render(
      <Provider store={storeRef.store}>
        <ResultsPage season={MOCKED_DATA} />
      </Provider>,
    );

    await act(async () => {
      storeRef.store.dispatch(resultsApi.endpoints.getResultsBySeason.initiate({ season: MOCKED_DATA.season }));
    });

    expect(screen.queryByAltText('loader')).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  test('renders with results without crashing', async () => {
    server.use(
      rest.get(`${API_URL}/${MOCKED_DATA.season}/results/1.json`, (_, res, ctx) =>
        res(ctx.status(200), ctx.json(getResultsBySeasonResponse)),
      ),
      rest.get(`${WIKI_API_URL}*`, (_, res, ctx) => res(ctx.status(500))),
    );

    const { container } = render(
      <Provider store={storeRef.store}>
        <ResultsPage season={MOCKED_DATA} />
      </Provider>,
    );

    await act(async () => {
      storeRef.store.dispatch(resultsApi.endpoints.getResultsBySeason.initiate({ season: MOCKED_DATA.season }));
    });

    expect(screen.queryByAltText('loader')).not.toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });
});
