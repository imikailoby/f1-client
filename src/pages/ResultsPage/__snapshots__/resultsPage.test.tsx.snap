// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ResultsPage /> renders with loading state without crashing 1`] = `
<div>
  <div
    class="results"
  >
    <h1>
      Season 
      <span>
        2005
      </span>
    </h1>
    <h3>
      World champion:
    </h3>
    <div
      class="champion"
    >
      <div
        class="driver"
      >
        <div
          class="avatar avatar"
        >
          <span>
            ...
          </span>
          <span
            class="number"
          >
            N/A
          </span>
        </div>
        <p
          class="name"
        >
          Nino
           
          <b>
            Farina
          </b>
        </p>
        <p
          class="car"
        >
          Alfa Romeo
        </p>
        <a
          class="link"
          href="http://en.wikipedia.org/wiki/Nino_Farina"
          rel="noreferrer"
          target="_blank"
        >
          Read more...
        </a>
      </div>
    </div>
    <h3>
      Results:
    </h3>
    <div
      class="loader"
    >
      <div
        class="loader"
      >
        <img
          alt="loader"
          src="svg>
"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`<ResultsPage /> renders with results without crashing 1`] = `
<div>
  <div
    class="results"
  >
    <h1>
      Season 
      <span>
        2005
      </span>
    </h1>
    <h3>
      World champion:
    </h3>
    <div
      class="champion"
    >
      <div
        class="driver"
      >
        <div
          class="avatar avatar"
        >
          <span>
            ...
          </span>
          <span
            class="number"
          >
            N/A
          </span>
        </div>
        <p
          class="name"
        >
          Nino
           
          <b>
            Farina
          </b>
        </p>
        <p
          class="car"
        >
          Alfa Romeo
        </p>
        <a
          class="link"
          href="http://en.wikipedia.org/wiki/Nino_Farina"
          rel="noreferrer"
          target="_blank"
        >
          Read more...
        </a>
      </div>
    </div>
    <h3>
      Results:
    </h3>
    <div
      class="table"
    >
      <div
        class="row"
      >
        <div>
          <p>
            Round
          </p>
        </div>
        <div>
          <p>
            Driver
          </p>
        </div>
        <div>
          <p>
            Race
          </p>
        </div>
      </div>
      <div
        class="row "
      >
        <div
          data-text="Round"
        >
          <p>
            1
          </p>
        </div>
        <div
          data-text="Driver"
        >
          <a
            href="http://en.wikipedia.org/wiki/Giancarlo_Fisichella"
            rel="noreferrer"
            target="_blank"
          >
            Giancarlo
             
            Fisichella
          </a>
          <span>
            Renault
          </span>
        </div>
        <div
          data-text="Race"
        >
          <a
            href="http://en.wikipedia.org/wiki/2005_Australian_Grand_Prix"
            rel="noreferrer"
            target="_blank"
          >
            Australian Grand Prix
          </a>
          <span>
            2005-03-06
          </span>
        </div>
      </div>
      <div
        class="row "
      >
        <div
          data-text="Round"
        >
          <p>
            2
          </p>
        </div>
        <div
          data-text="Driver"
        >
          <a
            href="http://en.wikipedia.org/wiki/Fernando_Alonso"
            rel="noreferrer"
            target="_blank"
          >
            Fernando
             
            Alonso
          </a>
          <span>
            Renault
          </span>
        </div>
        <div
          data-text="Race"
        >
          <a
            href="http://en.wikipedia.org/wiki/2005_Malaysian_Grand_Prix"
            rel="noreferrer"
            target="_blank"
          >
            Malaysian Grand Prix
          </a>
          <span>
            2005-03-20
          </span>
        </div>
      </div>
      <div
        class="row "
      >
        <div
          data-text="Round"
        >
          <p>
            3
          </p>
        </div>
        <div
          data-text="Driver"
        >
          <a
            href="http://en.wikipedia.org/wiki/Fernando_Alonso"
            rel="noreferrer"
            target="_blank"
          >
            Fernando
             
            Alonso
          </a>
          <span>
            Renault
          </span>
        </div>
        <div
          data-text="Race"
        >
          <a
            href="http://en.wikipedia.org/wiki/2005_Bahrain_Grand_Prix"
            rel="noreferrer"
            target="_blank"
          >
            Bahrain Grand Prix
          </a>
          <span>
            2005-04-03
          </span>
        </div>
      </div>
      <div
        class="row "
      >
        <div
          data-text="Round"
        >
          <p>
            4
          </p>
        </div>
        <div
          data-text="Driver"
        >
          <a
            href="http://en.wikipedia.org/wiki/Fernando_Alonso"
            rel="noreferrer"
            target="_blank"
          >
            Fernando
             
            Alonso
          </a>
          <span>
            Renault
          </span>
        </div>
        <div
          data-text="Race"
        >
          <a
            href="http://en.wikipedia.org/wiki/2005_San_Marino_Grand_Prix"
            rel="noreferrer"
            target="_blank"
          >
            San Marino Grand Prix
          </a>
          <span>
            2005-04-24
          </span>
        </div>
      </div>
      <div
        class="row "
      >
        <div
          data-text="Round"
        >
          <p>
            5
          </p>
        </div>
        <div
          data-text="Driver"
        >
          <a
            href="http://en.wikipedia.org/wiki/Kimi_R%C3%A4ikk%C3%B6nen"
            rel="noreferrer"
            target="_blank"
          >
            Kimi
             
            Räikkönen
          </a>
          <span>
            McLaren
          </span>
        </div>
        <div
          data-text="Race"
        >
          <a
            href="http://en.wikipedia.org/wiki/2005_Spanish_Grand_Prix"
            rel="noreferrer"
            target="_blank"
          >
            Spanish Grand Prix
          </a>
          <span>
            2005-05-08
          </span>
        </div>
      </div>
      <div
        class="row "
      >
        <div
          data-text="Round"
        >
          <p>
            6
          </p>
        </div>
        <div
          data-text="Driver"
        >
          <a
            href="http://en.wikipedia.org/wiki/Kimi_R%C3%A4ikk%C3%B6nen"
            rel="noreferrer"
            target="_blank"
          >
            Kimi
             
            Räikkönen
          </a>
          <span>
            McLaren
          </span>
        </div>
        <div
          data-text="Race"
        >
          <a
            href="http://en.wikipedia.org/wiki/2005_Monaco_Grand_Prix"
            rel="noreferrer"
            target="_blank"
          >
            Monaco Grand Prix
          </a>
          <span>
            2005-05-22
          </span>
        </div>
      </div>
      <div
        class="row "
      >
        <div
          data-text="Round"
        >
          <p>
            7
          </p>
        </div>
        <div
          data-text="Driver"
        >
          <a
            href="http://en.wikipedia.org/wiki/Fernando_Alonso"
            rel="noreferrer"
            target="_blank"
          >
            Fernando
             
            Alonso
          </a>
          <span>
            Renault
          </span>
        </div>
        <div
          data-text="Race"
        >
          <a
            href="http://en.wikipedia.org/wiki/2005_European_Grand_Prix"
            rel="noreferrer"
            target="_blank"
          >
            European Grand Prix
          </a>
          <span>
            2005-05-29
          </span>
        </div>
      </div>
      <div
        class="row "
      >
        <div
          data-text="Round"
        >
          <p>
            8
          </p>
        </div>
        <div
          data-text="Driver"
        >
          <a
            href="http://en.wikipedia.org/wiki/Kimi_R%C3%A4ikk%C3%B6nen"
            rel="noreferrer"
            target="_blank"
          >
            Kimi
             
            Räikkönen
          </a>
          <span>
            McLaren
          </span>
        </div>
        <div
          data-text="Race"
        >
          <a
            href="http://en.wikipedia.org/wiki/2005_Canadian_Grand_Prix"
            rel="noreferrer"
            target="_blank"
          >
            Canadian Grand Prix
          </a>
          <span>
            2005-06-12
          </span>
        </div>
      </div>
      <div
        class="row "
      >
        <div
          data-text="Round"
        >
          <p>
            9
          </p>
        </div>
        <div
          data-text="Driver"
        >
          <a
            href="http://en.wikipedia.org/wiki/Michael_Schumacher"
            rel="noreferrer"
            target="_blank"
          >
            Michael
             
            Schumacher
          </a>
          <span>
            Ferrari
          </span>
        </div>
        <div
          data-text="Race"
        >
          <a
            href="http://en.wikipedia.org/wiki/2005_United_States_Grand_Prix"
            rel="noreferrer"
            target="_blank"
          >
            United States Grand Prix
          </a>
          <span>
            2005-06-19
          </span>
        </div>
      </div>
      <div
        class="row "
      >
        <div
          data-text="Round"
        >
          <p>
            10
          </p>
        </div>
        <div
          data-text="Driver"
        >
          <a
            href="http://en.wikipedia.org/wiki/Fernando_Alonso"
            rel="noreferrer"
            target="_blank"
          >
            Fernando
             
            Alonso
          </a>
          <span>
            Renault
          </span>
        </div>
        <div
          data-text="Race"
        >
          <a
            href="http://en.wikipedia.org/wiki/2005_French_Grand_Prix"
            rel="noreferrer"
            target="_blank"
          >
            French Grand Prix
          </a>
          <span>
            2005-07-03
          </span>
        </div>
      </div>
      <div
        class="row "
      >
        <div
          data-text="Round"
        >
          <p>
            11
          </p>
        </div>
        <div
          data-text="Driver"
        >
          <a
            href="http://en.wikipedia.org/wiki/Juan_Pablo_Montoya"
            rel="noreferrer"
            target="_blank"
          >
            Juan
             
            Pablo Montoya
          </a>
          <span>
            McLaren
          </span>
        </div>
        <div
          data-text="Race"
        >
          <a
            href="http://en.wikipedia.org/wiki/2005_British_Grand_Prix"
            rel="noreferrer"
            target="_blank"
          >
            British Grand Prix
          </a>
          <span>
            2005-07-10
          </span>
        </div>
      </div>
      <div
        class="row "
      >
        <div
          data-text="Round"
        >
          <p>
            12
          </p>
        </div>
        <div
          data-text="Driver"
        >
          <a
            href="http://en.wikipedia.org/wiki/Fernando_Alonso"
            rel="noreferrer"
            target="_blank"
          >
            Fernando
             
            Alonso
          </a>
          <span>
            Renault
          </span>
        </div>
        <div
          data-text="Race"
        >
          <a
            href="http://en.wikipedia.org/wiki/2005_German_Grand_Prix"
            rel="noreferrer"
            target="_blank"
          >
            German Grand Prix
          </a>
          <span>
            2005-07-24
          </span>
        </div>
      </div>
      <div
        class="row "
      >
        <div
          data-text="Round"
        >
          <p>
            13
          </p>
        </div>
        <div
          data-text="Driver"
        >
          <a
            href="http://en.wikipedia.org/wiki/Kimi_R%C3%A4ikk%C3%B6nen"
            rel="noreferrer"
            target="_blank"
          >
            Kimi
             
            Räikkönen
          </a>
          <span>
            McLaren
          </span>
        </div>
        <div
          data-text="Race"
        >
          <a
            href="http://en.wikipedia.org/wiki/2005_Hungarian_Grand_Prix"
            rel="noreferrer"
            target="_blank"
          >
            Hungarian Grand Prix
          </a>
          <span>
            2005-07-31
          </span>
        </div>
      </div>
      <div
        class="row "
      >
        <div
          data-text="Round"
        >
          <p>
            14
          </p>
        </div>
        <div
          data-text="Driver"
        >
          <a
            href="http://en.wikipedia.org/wiki/Kimi_R%C3%A4ikk%C3%B6nen"
            rel="noreferrer"
            target="_blank"
          >
            Kimi
             
            Räikkönen
          </a>
          <span>
            McLaren
          </span>
        </div>
        <div
          data-text="Race"
        >
          <a
            href="http://en.wikipedia.org/wiki/2005_Turkish_Grand_Prix"
            rel="noreferrer"
            target="_blank"
          >
            Turkish Grand Prix
          </a>
          <span>
            2005-08-21
          </span>
        </div>
      </div>
      <div
        class="row "
      >
        <div
          data-text="Round"
        >
          <p>
            15
          </p>
        </div>
        <div
          data-text="Driver"
        >
          <a
            href="http://en.wikipedia.org/wiki/Juan_Pablo_Montoya"
            rel="noreferrer"
            target="_blank"
          >
            Juan
             
            Pablo Montoya
          </a>
          <span>
            McLaren
          </span>
        </div>
        <div
          data-text="Race"
        >
          <a
            href="http://en.wikipedia.org/wiki/2005_Italian_Grand_Prix"
            rel="noreferrer"
            target="_blank"
          >
            Italian Grand Prix
          </a>
          <span>
            2005-09-04
          </span>
        </div>
      </div>
      <div
        class="row "
      >
        <div
          data-text="Round"
        >
          <p>
            16
          </p>
        </div>
        <div
          data-text="Driver"
        >
          <a
            href="http://en.wikipedia.org/wiki/Kimi_R%C3%A4ikk%C3%B6nen"
            rel="noreferrer"
            target="_blank"
          >
            Kimi
             
            Räikkönen
          </a>
          <span>
            McLaren
          </span>
        </div>
        <div
          data-text="Race"
        >
          <a
            href="http://en.wikipedia.org/wiki/2005_Belgian_Grand_Prix"
            rel="noreferrer"
            target="_blank"
          >
            Belgian Grand Prix
          </a>
          <span>
            2005-09-11
          </span>
        </div>
      </div>
      <div
        class="row "
      >
        <div
          data-text="Round"
        >
          <p>
            17
          </p>
        </div>
        <div
          data-text="Driver"
        >
          <a
            href="http://en.wikipedia.org/wiki/Juan_Pablo_Montoya"
            rel="noreferrer"
            target="_blank"
          >
            Juan
             
            Pablo Montoya
          </a>
          <span>
            McLaren
          </span>
        </div>
        <div
          data-text="Race"
        >
          <a
            href="http://en.wikipedia.org/wiki/2005_Brazilian_Grand_Prix"
            rel="noreferrer"
            target="_blank"
          >
            Brazilian Grand Prix
          </a>
          <span>
            2005-09-25
          </span>
        </div>
      </div>
      <div
        class="row "
      >
        <div
          data-text="Round"
        >
          <p>
            18
          </p>
        </div>
        <div
          data-text="Driver"
        >
          <a
            href="http://en.wikipedia.org/wiki/Kimi_R%C3%A4ikk%C3%B6nen"
            rel="noreferrer"
            target="_blank"
          >
            Kimi
             
            Räikkönen
          </a>
          <span>
            McLaren
          </span>
        </div>
        <div
          data-text="Race"
        >
          <a
            href="http://en.wikipedia.org/wiki/2005_Japanese_Grand_Prix"
            rel="noreferrer"
            target="_blank"
          >
            Japanese Grand Prix
          </a>
          <span>
            2005-10-09
          </span>
        </div>
      </div>
      <div
        class="row "
      >
        <div
          data-text="Round"
        >
          <p>
            19
          </p>
        </div>
        <div
          data-text="Driver"
        >
          <a
            href="http://en.wikipedia.org/wiki/Fernando_Alonso"
            rel="noreferrer"
            target="_blank"
          >
            Fernando
             
            Alonso
          </a>
          <span>
            Renault
          </span>
        </div>
        <div
          data-text="Race"
        >
          <a
            href="http://en.wikipedia.org/wiki/2005_Chinese_Grand_Prix"
            rel="noreferrer"
            target="_blank"
          >
            Chinese Grand Prix
          </a>
          <span>
            2005-10-16
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;
